### ISSUE TITLE
App may crash on wake from sleep or display connect/disconnect

### ISSUE THREAD

**Issue by Anywhere-Music-Player** [MonitorControl/MonitorControl#1648]
### Before opening the issue, have you...?

- [X] Searched for existing issues
- [X] Looked through [the wiki](https://github.com/MonitorControl/MonitorControl/wiki)
- [X] Updated MonitorControl to the latest version (if applicable)

### Describe the bug

After sleep, then wake up when I press F1/F2 brightness just drops to like 0% and stopped to respond, no changes at all.
Have to restart app to make it work. Well, I mean any way of changing brightness is not working, with sliders for example.

![Screenshot 2](https://github.com/user-attachments/assets/7e6187df-c930-4658-a0f7-fad9d05f6a19)

### Steps to reproduce

1. Sleep
2. Wake up
3. F1/F2

### Expected behavior

Brightness should always work. 

### Environment Information (please complete the following information)

```markdown
- macOS version: 15.0.1 (24A348)
- Mac model: MacBook Pro M1 14"
- MonitorControl version: 4.3.3
- Monitor(s): M28U
- Apple Silicon/M1 (yes or no): Yes
```

**Comment by waydabber** [comment]
Can't seem to reproduce this. Let's see if there are at least some other users reporting this issue, otherwise it might be a configuration specific problem or interference with some other app possibly difficult to track down.

**Comment by Volodymyr-13** [comment]
> Can't seem to reproduce this. Let's see if there are at least some other users reporting this issue, otherwise it might be a configuration specific problem or interference with some other app possibly difficult to track down.

Oh, sorry, forgot to mention, chance of this like happens only ~once per last 3 days so far. 

This bug was also with previous version, but less often I'd say.. so I don't think it just something new. 
Will update if this happens again, but I'm 100% sure that this is some bug. At the same time I've running my old app https://github.com/Volodymyr-13/NativeDisplayBrightness and when Monitor Control doesn't work, NativeDisplayBrightness works.

**Comment by waydabber** [comment]
No problem, thanks for the info. I did not make any changes in the app logic at all, just focused on compatibility updates. So it makes sense if this happened with previous versions as well.

**Comment by lerschiboi** [comment]
Not sure if it's the same problem, but since Sequoia, Monitor Control is always closed when I wake my computer from sleep. I have to start it again and then everything works fine again. 🤷

**Comment by Volodymyr-13** [comment]
> Not sure if it's the same problem, but since Sequoia, Monitor Control is always closed when I wake my computer from sleep. I have to start it again and then everything works fine again. 🤷

@waydabber  Oh yeah, me too, it's not 100% chance, but I'd say it's like 30% of chance at least. 
After wake - app crashed. Had to open it, yea.

**Comment by Volodymyr-13** [comment]
> Version 4.2.0 7048
> 
> True, just noticed that it's not after every wake, but happens quite a lot

And actually this happens to me for like at least 6+ months so this is an old bug.

**Comment by lerschiboi** [comment]
> > Version 4.2.0 7048
> > True, just noticed that it's not after every wake, but happens quite a lot
> 
> And actually this happens to me for like at least 6+ months so this is an old bug.

Well, I was rocking MacOS 14.4 for quite a long time before updating to 15.0, and it was only then that I started noticing the crashes. But maybe they were introduced with an earlier MacOS version that I didn't update to

**Comment by Volodymyr-13** [comment]
> (note: 4.2.0 is not the latest version)

And this and bug I reported is like the same, after wake was on any of those versions and I’m always updating to the very latest. So old version or new, situation remains the same…

**Comment by onekiloparsec** [comment]
I confirm I have the same behavior. Back from sleep, the app is not running anymore, as far as I can tell, every time (or close). I'am running macOS Sequioa `15.0.1` and MonitorControl `4.2.0 Build 7048`. With Spotlight, it's quite easy to relaunch, but it would be nice to not have to. :-)

**Comment by waydabber** [comment]
Note: 4.2.0 is not properly compatible with Sequoia, make sure you use the latest version. But other than that the issue might be valid on some configs as others pointed out. If I can get an ips crash log (of 4.3.3), I can check where the app crashed and probably fix it.

**Comment by onekiloparsec** [comment]
Oh thanks, I will update. Note that I didn't know 4.3.3 was available, because it didn't appear when checking for updates... 
![Screenshot 2024-10-12 at 09 37 04](https://github.com/user-attachments/assets/55732952-d751-4005-acd7-7e16665db921)

**Comment by natapatpimsang** [comment]
I faced same issue, I use MC 4.3.3 and my macos is 15.0.1

**Comment by waydabber** [comment]
Again: if I can get an ips crash log (of 4.3.3), I can check where the app crashed and probably fix it. :)

**Comment by Volodymyr-13** [comment]
> Again: if I can get an ips crash log (of 4.3.3), I can check where the app crashed and probably fix it. :)

For me it’s not crashing in all last cases.. but just stopping to work.. so maybe need some debug log or so?

**Comment by waydabber** [comment]
Oh, I see. Two issues seem to be conflated here.

**Comment by Volodymyr-13** [comment]
> Oh, I see. Two issues seem to be conflated here.

Oh sorry, I found it finally! it crashes for me too: 
[MonitorControl-2024-10-21-004416.ips.zip](https://github.com/user-attachments/files/17474085/MonitorControl-2024-10-21-004416.ips.zip)

**Comment by waydabber** [comment]
Thanks @Volodymyr-13!

**Comment by waydabber** [comment]
I analyzed the crash log, the crash occurs because of a rather unprofessional way of handling the contents of the percentage output when it is shown. The issue will probably go away showing percentages is turned off (unless there is an other crash hidden by this one). I fixed the issue so this should work better in the next release.

**Comment by Volodymyr-13** [comment]
@waydabber thanks, for now I'll try to disable % and test this until release.

**Comment by Volodymyr-13** [comment]
@waydabber @waydabber Hi, so all this time after I disabled % in the menu - the app worked well after sleep. 
But today this happened again:
 - after waking from sleep 
 - I tried to change brightness and it immediately went to 0%
 - after that I no longer can change brightness for my external monitor, it always stays 0%, sliders/hotkeys don't work for the external monitor
 - when I open app settings - it's like it opens and then disappears/goes out of focus, first I thought it's crashed but it's not
 - relaunch the app as usual - not fixed the problem, brightness stays 0% all the time
 - relaunch the app a second time, try opening settings - same problem like it opens/hides/out of focus in the background window under other Mac windows..
 - changed the brightness of the monitor manually and after that the app started to work 
 - no crashes found in the Console afterwards

![Screenshot 2 Medium](https://github.com/user-attachments/assets/9a1c221b-0d73-40c1-b110-58936c9533ef)
![Screenshot 1 Medium](https://github.com/user-attachments/assets/c7d9e7b1-e9b2-4ed2-b914-dca122f82db9)
![Screenshot Medium](https://github.com/user-attachments/assets/75df0c68-42f8-4694-92e4-2c9b243047a0)

**Comment by waydabber** [comment]
I don't think this is the same issue as it is not an app crash.

**Comment by Volodymyr-13** [comment]
> I don't think this is the same issue as it is not an app crash.

I think the same, but these are like very related things, because they occur at this very moment, which was originally described in this bug. Now it is difficult to investigate this, maybe there is some log that the monitor keeps... to see what is happening there..? 

Hmm, this now happens very often for me now.. looks like I need to reboot macOS to "fix" this... will not do that yet..

**Comment by Volodymyr-13** [comment]
@waydabber, for the first time ever, this issue has occurred after a long time since I woke up my MacBook.

Randomly, whenever I adjust the brightness (which I used to do and it worked well), it drops to 0% and stops working immediately. Only restarting the app helps. Interestingly, if I close the app and immediately start it again, it won't run. I need to take a delay of about 2-3 seconds before starting it again, and then it starts and works.

**Comment by jmedina21** [comment]
I am experiencing a similar issue, when I mute and unmute it stays muted, I have to restart the app, and the monitor restarts automatically.
![image](https://github.com/user-attachments/assets/b9458446-7d77-4add-bb56-5b000ef8d2f2)

Steps to reproduce
Mute
Unmute

Expected behavior
It should unmute

Anything else?
No response

Environment Information (please complete the following information)
- macOS version: 15.1.1 (24B2091)
- Mac model: Mac Mini M4 Pro
- MonitorControl version: 4.3.3
- Monitor(s): Samsung Odyssey g5
- Apple Silicon/M1 (yes or no): Yes

**Comment by Volodymyr-13** [comment]
> I am experiencing a similar issue, when I mute and unmute it stays muted, I have to restart the app, and the monitor restarts automatically. ![image](https://private-user-images.githubusercontent.com/22413563/388589606-b9458446-7d77-4add-bb56-5b000ef8d2f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyMDUyNDIsIm5iZiI6MTczMjIwNDk0MiwicGF0aCI6Ii8yMjQxMzU2My8zODg1ODk2MDYtYjk0NTg0NDYtN2Q3Ny00YWRkLWJiNTYtNWIwMDBlZjhkMmYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIxVDE2MDIyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI2ZWRhMzhhNWRiMjNiZDcxYTczOGMzZGY1ZWY5NjA4OTExZGU2YjU1NGFhYTVjODQ0MzkzOTMwOWI0ZTBkM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iLdBzuf9KOjKdcZGvc0LJp-MlDODor10UqLSrrQ4m5c)
> 
> Steps to reproduce Mute Unmute
> 
> Expected behavior It should unmute
> 
> Anything else? No response
> 
> Environment Information (please complete the following information)
> 
> * macOS version: 15.1.1 (24B2091)
> * Mac model: Mac Mini M4 Pro
> * MonitorControl version: 4.3.3
> * Monitor(s): Samsung Odyssey g5
> * Apple Silicon/M1 (yes or no): Yes

Thats kind of different issue, I have it too, but forgot to report it, since I almost never used Mute, just accidentally pressing it). 
I think you better to create a new bug report..

**Comment by Volodymyr-13** [comment]
@waydabber crashed today again(with disabled % all this time).

[MonitorControl-2024-11-22-130821.ips.zip](https://github.com/user-attachments/files/17869579/MonitorControl-2024-11-22-130821.ips.zip)

**Comment by waydabber** [comment]
This is a different issue. The crash log comes from a place where the system tries to convert a value (probably brightness) to a DDC output based on the scaling curve and min/max DDC values. Is there any special configuration/change made in the Advanced section for the display? I can add some sanitization and safety checks to the variables involved in general but without the exact circumstances I have no way of knowing what exactly is the root cause.

For future reference, here is the symbolicated ips. [MonitorControl-2024-11-22-130821_symbolicated.ips.zip](https://github.com/user-attachments/files/17869802/MonitorControl-2024-11-22-130821_symbolicated.ips.zip)

**Comment by Volodymyr-13** [comment]
> This is a different issue. The crash log comes from a place where the system tries to convert a value (probably brightness) to a DDC output based on the scaling curve and min/max DDC values. Is there any special configuration/change made in the Advanced section for the display? I can add some sanitization and safety checks to the variables involved in general but without the exact circumstances I have no way of knowing what exactly is the root cause.
> 
> For future reference, here is the symbolicated ips. [MonitorControl-2024-11-22-130821_symbolicated.ips.zip](https://github.com/user-attachments/files/17869802/MonitorControl-2024-11-22-130821_symbolicated.ips.zip)

Nope, I never opened advanced settings until now, and this bug just randomly occurred after waking the system from sleep. Only one thing changed recently: due to the russian orcs attacking with ballistic rockets our country a couple of days ago, we now have blackouts. This left the monitor without power while the MacBook continued on battery. Maybe the system couldn’t read the monitor’s values properly then. 

![Screenshot 2 Medium](https://github.com/user-attachments/assets/c368c8d3-2588-4da3-b97f-ce46d7afa59a)
![Screenshot 3 Medium](https://github.com/user-attachments/assets/bf49e5d9-43b5-4f06-b3c7-6d5b3b9d6a2e)

**Comment by waydabber** [comment]
Indeed, this is a different crash (apparently a wrong value was saved to preferences and the app has trouble converting it to a DDC value. Not sure why and how though.

**Comment by Volodymyr-13** [comment]
> Indeed, this is a different crash (apparently a wrong value was saved to preferences and the app has trouble converting it to a DDC value. Not sure why and how though.

Situation was - MacBook was on power and monitor too, the power down, MacBook awake(just show Lock Screen) then when power came back, only after that I unlocked MacBook so to continue working with external monitor on power.

**Comment by waydabber** [comment]
No problem, I am not sure exactly why the situation arises, but clearly a wrong value is being used. I just need to add some protections and clamping to make sure that particular line does not fail.

**Comment by waydabber** [comment]
https://github.com/MonitorControl/MonitorControl/issues/1706

**Comment by Volodymyr-13** [comment]
@waydabber so the issue continues, awake from sleep = crash. 
Try start app again - crash, start again - changing brightness doesn't work, start again - crash, start again - now all works. 

% of brightness never enabled now for all this time. 

[MonitorControl-2025-01-24-115914.zip](https://github.com/user-attachments/files/18533759/MonitorControl-2025-01-24-115914.zip)

![Image](https://github.com/user-attachments/assets/99dccf70-9cb4-446c-aaba-3733f36acb7c)

**Comment by waydabber** [comment]
Yep, this seems to be the same (not % related) crash as above.

**Comment by mateusz-plociennik** [comment]
I have built `4.3.3` from sources and caught a fatal error: 
`MonitorControl/MenuHandler.swift:50: Fatal error: Unexpectedly found nil while unwrapping an Optional value`
<img width="2560" height="1415" alt="Image" src="https://github.com/user-attachments/assets/1c6be7ad-e72f-406c-a010-621c5366fe38" />
